### PR TITLE
Add docusaurus documentation site

### PR DESCRIPTION
**Pull request by silky**
Hello!

Lovely software :)

Here is a small PR that adds a docusaurus website (locally only) that displays a selection of relevant documentation coming from the output of `b6-api --docs --functions`.

Here's how it looks:

image

~I also took the liberty to do a minor improvement to how the `Collections` are printed; instead of `FeatureIDFeatureIDCollection` it now emits `Collection[FeatureID, FeatureID]`.~

Finally, I've included some Makefile targets to generate this; so you should be able to just run `make docs` and it will (hopefully!) work.

Let me know what you think! Happy to make any changes based on feedback.

Open questions:

Technical:
- [ ] `Query` comes out twice; why?
- I've done a "best guess" (i.e. horrible guess) at the python name of the functions; (`"-" -> "_")`; i think it's fine enough.
- Obviously, it's a bit intermingled, as the types aren't python-specific; but perhaps it serves as a good starting point
- Sometimes, the links to other types don't work prefectly (namely, in the `Collection[a,b]` case; but in part that's because I'm not even writing the `"Collections"` field of the `b6-api` output (as it seems a bit redundant when we are doing the names differently.)

Usability:
- No attempt is made to follow any style guide on the docs; this could come subsequently if we are happy with it. The website looks bare; but all we care about here is these items being documented
- There's no gh-pages deployment tasks; this could be added at some point
- [x] ~I've made a strange choice to document the return type as `:: A` for some type `A` so it's easy to search for functions that return the same thing. This is Haskell syntax. Please forgive me or suggest an alternative :)~ Using Python `->` style.
- [x] `I write `b6.` in front of the (pythonised) names, again so it's easy to search for specific function invocations.` Can now search for `def ...`
- [ ] It would benefit the rendering of the go docstrings if they used markdown, given that this is how we interpret them. I'm not sure if this is acceptable or incompatibile with, say, `godoc` (or `go doc` ...?)

Trivia:
- I did actually try getting (similar) documentation to this with `godoc`; but in fact I didn't see these particular functions documented at all by a naive invocation of that tool. Maybe that's possible somehow and is a plausible alternative?

### CODE REVIEW

**Comment by silky** [comment]
Perhaps, upon review, it would be better to leave things a bit more technically correct, but less convenient, and refer to this as the "protobuf" API, with the specific function names, and then give some tips about how to interpret this in Python?

Or perhaps we could just guess/compute the python function name (and module location) and display that alongside the other information?

I'm open-minded.

**Review by andreweland — APPROVED** [review]
Looks good - thanks for starting this!

With regards to formatting the output, I think we should stick to something that looks like Python type hints, since that's the primary way people interact with b6 at the moment - ie using `->` for the return type etc.

In terms of guessing the Python name, your rules are pretty much those used by the code (`escape_name` in python/diagonal_b6/generate_api.py). It's not great duplicating that logic though. I wouldn't be averse to moving it into go, and having b6-api output both a generic name and a Python specific name for things. Long term, we hope to support other languages (specifically JS sooner rather than later), and so we'd want the ability to escape for that too. At that point, we'd probably want an output documentation format that was generic, but with eg specific examples for each language in a kind of tab-ui type thing.

These can all be for future PRs though.

File: docs/docusaurus.config.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
+import {themes as prismThemes} from 'prism-react-renderer';
+import type {Config} from '@docusaurus/types';
+import type * as Preset from '@docusaurus/preset-classic';
+
+const config: Config = {
+  title: 'diagonal b6',
+  // tagline: 'Dinosaurs are cool',
```
[the remark this answers is not in the document]
Comment (silky) [reply docs/docusaurus.config.ts:7]: Good point, will do.

File: src/diagonal.works/b6/cmd/b6-api/b6-api.go [hunk 2]
```diff
@@ -77,7 +77,7 @@ func isBuiltin(t reflect.Type) bool {
 func collectionForType(t reflect.Type) Collection {
 	if t == UntypedCollectionType {
 		return Collection{
-			Name:  "AnyAnyCollection",
+			Name:  "Collection[Any,Any]",
```
Comment (silky) [inline src/diagonal.works/b6/cmd/b6-api/b6-api.go:80]: Upon reflection I wonder if this breaks the Python codegen; I'll double check.

Comment (silky) [reply src/diagonal.works/b6/cmd/b6-api/b6-api.go:80]: It does. I'll revert.

**Comment by silky** [comment]
This is ready now; no more breaking changes, and I updated the screenshot to show how it looks presently (made some minor changes).

If you can approve the PR actions kick-off it should be good to go!

Thanks @andreweland :)

**Comment by silky** [comment]
> At that point, we'd probably want an output documentation format that was generic, but with eg specific examples for 
> each language in a kind of tab-ui type thing.

Yes; makes a lot of sense, and indeed I think docusaurus is pretty good for this kind of thing; as you say will just need a minor re-org in how we do all the generation; but that should be fine.

**Comment by andreweland** [comment]
The way we have things setup now, the status checks will only run from PRs that originate from the origin repo (because they run on GCP, and the security issues would be tedious to work through), so I ran the tests locally - all was good, so I'll merge and skip the checks.

**Pull request merged**